l2-contracts/contracts/bridge/L2NativeTokenVault.sol

```diff
@@ -0,0 +1,187 @@
+    /// @dev Disable the initialization to prevent Parity hack.
+    constructor() {
+        _disableInitializers();
+    }
+
+    /// @dev Sets the L1ERC20Bridge contract address. Should be called only once.
+    function setSharedBridge(IL2SharedBridge _sharedBridge) external onlyOwner {
+        if (address(l2Bridge) != address(0)) {
+            // "SD: shared bridge already set";
+            revert AddressMismatch(address(0), address(l2Bridge));
+        }
+        if (address(_sharedBridge) == address(0)) {
+            // "SD: shared bridge 0");
+            revert EmptyAddress();
+        }
+
+        l2Bridge = _sharedBridge;
+    }
+
+    /// @notice Initializes the bridge contract for later use. Expected to be used in the proxy.
+    /// @param _l2TokenProxyBytecodeHash The bytecode hash of the proxy for tokens deployed by the bridge.
+    /// @param _aliasedOwner The address of the governor contract.
+    /// @param _contractsDeployedAlready Ensures beacon proxy for standard ERC20 has not been deployed
+    function initialize(
+        bytes32 _l2TokenProxyBytecodeHash,
+        address _aliasedOwner,
+        bool _contractsDeployedAlready
+    ) external reinitializer(2) {
+        if (_l2TokenProxyBytecodeHash == bytes32(0)) {
+            revert EmptyBytes32();
+        }
+        if (_aliasedOwner == address(0)) {
+            revert EmptyAddress();
+        }
+
+        if (!_contractsDeployedAlready) {
+            address l2StandardToken = address(new L2StandardERC20{salt: bytes32(0)}());
+            l2TokenBeacon = new UpgradeableBeacon{salt: bytes32(0)}(l2StandardToken);
+            l2TokenProxyBytecodeHash = _l2TokenProxyBytecodeHash;
+            l2TokenBeacon.transferOwnership(_aliasedOwner);
+        }
+
+        _transferOwnership(_aliasedOwner);
+    }
+
+    function setL2TokenBeacon(address _l2TokenBeacon, bytes32 _l2TokenProxyBytecodeHash) external onlyOwner {
+        l2TokenBeacon = UpgradeableBeacon(_l2TokenBeacon);
+        l2TokenProxyBytecodeHash = _l2TokenProxyBytecodeHash;
+        emit L2TokenBeaconUpdated(_l2TokenBeacon, _l2TokenProxyBytecodeHash);
+    }
+
+    function bridgeMint(uint256 _chainId, bytes32 _assetId, bytes calldata _data) external payable override {
+        address token = tokenAddress[_assetId];
+        (address _l1Sender, uint256 _amount, address _l2Receiver, bytes memory erc20Data, address originToken) = abi
+            .decode(_data, (address, uint256, address, bytes, address));
+        address expectedToken = l2TokenAddress(originToken);
+        if (token == address(0)) {
+            bytes32 expectedAssetId = keccak256(
+                abi.encode(_chainId, NATIVE_TOKEN_VAULT_VIRTUAL_ADDRESS, bytes32(uint256(uint160(originToken))))
+            );
+            if (_assetId != expectedAssetId) {
+                // Make sure that a NativeTokenVault sent the message
+                revert Bytes32Mismatch(_assetId, expectedAssetId);
```

Comment:
A better named error could be used, e.g. assetIdMismatch or something

Reply:
This is how Zach in protocolDefence did it, i.e. it is named after the type.